### DISCUSSION TITLE
jsonable_encoder can't roundtrip properly with ConstrainedDecimal

### DISCUSSION THREAD

**Discussion by Hultner** [fastapi/fastapi#6990] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

I wrote the following test/script to showcase the problem, I have a ConstrainedDecimal which represents a `Numeric(22,0)`.

**test_decimal_bug.py:**
```python
from pydantic import ConstrainedDecimal, BaseModel
from decimal import Decimal
from fastapi.encoders import jsonable_encoder


class Id(ConstrainedDecimal):
    max_digits = 22
    decimal_places = 0
    ge = 0


ObjId = Id


class Obj(BaseModel):
    id: ObjId
    name: str


def test_con_decimal_encode() -> None:
    test_obj = Obj(id=1, name="Test Obj")
    cycled_obj = Obj.parse_obj(jsonable_encoder(test_obj))
    assert test_obj == cycled_obj


def test_con_decimal_encode_custom_encoder() -> None:
    test_obj = Obj(id=1, name="Test Obj")
    cycled_obj = Obj.parse_obj(jsonable_encoder(test_obj, custom_encoder={Id: int}))
    assert test_obj == cycled_obj


def test_con_decimal_encode_custom_encoder_2() -> None:
    test_obj = Obj(id=1, name="Test Obj")
    cycled_obj = Obj.parse_obj(jsonable_encoder(test_obj, custom_encoder={ObjId: int}))
    assert test_obj == cycled_obj


if __name__ == "__main__":
    from pprint import pprint

    print_obj = Obj(id=1, name="Test Obj")
    jsonable_obj = jsonable_encoder(print_obj, custom_encoder={Id: int})
    pprint(jsonable_obj)
```

Result from running it as a script
```sh
$ python test_decimal_bug.py
{'id': 1.0, 'name': 'Test Obj'}

```

And with pytest -v
```sh
$ pytest -v test_decimal_bug.py
========================================================== test session starts ===========================================================
platform darwin -- Python 3.9.0, pytest-6.2.2, py-1.10.0, pluggy-0.13.1 -- /Users/hultner/Library/Caches/pypoetry/virtualenvs/hantera-backend-5YPPF0bI-py3.9/bin/python
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/Users/hultner/dev/hantera/hantera_backend/.hypothesis/examples')
rootdir: /Users/hultner/dev/hantera/hantera_backend, configfile: pyproject.toml
plugins: cov-2.11.1, subtests-0.4.0, testmon-1.0.3, xdist-2.2.0, forked-1.3.0, schemathesis-2.8.5, hypothesis-5.49.0
collected 3 items

test_decimal_bug.py::test_con_decimal_encode FAILED                                                                                [ 33%]
test_decimal_bug.py::test_con_decimal_encode_custom_encoder FAILED                                                                 [ 66%]
test_decimal_bug.py::test_con_decimal_encode_custom_encoder_2 FAILED                                                               [100%]

================================================================ FAILURES ================================================================
________________________________________________________ test_con_decimal_encode _________________________________________________________

    def test_con_decimal_encode() -> None:
        test_obj = Obj(id=1, name="Test Obj")
>       cycled_obj = Obj.parse_obj(jsonable_encoder(test_obj))

test_decimal_bug.py:22:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pydantic/main.py:520: in pydantic.main.BaseModel.parse_obj
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   pydantic.error_wrappers.ValidationError: 1 validation error for Obj
E   id
E     ensure that there are no more than 0 decimal places (type=value_error.decimal.max_places; decimal_places=0)

pydantic/main.py:362: ValidationError
_________________________________________________ test_con_decimal_encode_custom_encoder _________________________________________________

    def test_con_decimal_encode_custom_encoder() -> None:
        test_obj = Obj(id=1, name="Test Obj")
>       cycled_obj = Obj.parse_obj(jsonable_encoder(test_obj, custom_encoder={Id: int}))

test_decimal_bug.py:28:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pydantic/main.py:520: in pydantic.main.BaseModel.parse_obj
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   pydantic.error_wrappers.ValidationError: 1 validation error for Obj
E   id
E     ensure that there are no more than 0 decimal places (type=value_error.decimal.max_places; decimal_places=0)

pydantic/main.py:362: ValidationError
________________________________________________ test_con_decimal_encode_custom_encoder_2 ________________________________________________

    def test_con_decimal_encode_custom_encoder_2() -> None:
        test_obj = Obj(id=1, name="Test Obj")
>       cycled_obj = Obj.parse_obj(jsonable_encoder(test_obj, custom_encoder={ObjId: int}))

test_decimal_bug.py:34:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pydantic/main.py:520: in pydantic.main.BaseModel.parse_obj
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   pydantic.error_wrappers.ValidationError: 1 validation error for Obj
E   id
E     ensure that there are no more than 0 decimal places (type=value_error.decimal.max_places; decimal_places=0)

pydantic/main.py:362: ValidationError
======================================================== short test summary info =========================================================
FAILED test_decimal_bug.py::test_con_decimal_encode - pydantic.error_wrappers.ValidationError: 1 validation error for Obj
FAILED test_decimal_bug.py::test_con_decimal_encode_custom_encoder - pydantic.error_wrappers.ValidationError: 1 validation error for Obj
FAILED test_decimal_bug.py::test_con_decimal_encode_custom_encoder_2 - pydantic.error_wrappers.ValidationError: 1 validation error for Obj
=========================================================== 3 failed in 0.22s ============================================================

```

### Description

* Run the test script with `python test_decimal_bug.py`
* It returns a JSON with `{'id': 1.0, 'name': 'Test Obj'}`.
* But I expected it to return `{'id': 1, 'name': 'Test Obj'}`.

### Environment

* OS: [e.g. Linux / Windows / macOS]: macOS & Linux
* FastAPI Version [e.g. 0.3.0]: 0.62.0
* Python version: 3.9.0

### Additional context
When I return the pydantic objects through an endpoint, the IDs show up correctly without the trailing invalid decimal, which further confuses me as the documentation states that this would use the very same encoder.

As you can see in the tests I also tried to register a custom encoder for Ids (both the base class and alias) but this doesn't help.

**Comment by Hultner** [comment]
Can't change the label but this is a problem rather than a question. 

However a question would be if you can suggest a suitable workaround for now?

**Comment by Hultner** [comment]
After looking through the source code of FastAPI I can see that the encoders themself are provided by pydantic. 
I have posted a patch to pydantic which fixes that issue there. 

This should hopefully resolve itself here if my patch is merged in pydantic. 

**Relevant issue and patch:** https://github.com/samuelcolvin/pydantic/pull/2294 https://github.com/samuelcolvin/pydantic/pull/2293

**Comment by Hultner** [comment]
The patch for this bug is now merged via: https://github.com/samuelcolvin/pydantic/pull/2294
And will thus be fixed with the next release of pydantic.

However it's still a second aspect of this where I couldn't provide a custom encoder for my type without overwriting the super type. Not sure if this would belong here or in pydantic (or both). But it would make sense for me that I should be able to provide a custom encoder for my Id type, and that encoder being used for those properties.

**Comment by Hultner** [comment, marked as answer]
Forgot to close this. If anyone else runs into this problem, make sure to update pydantic and it should be fine.

**Comment by tiangolo** [comment]
Thanks for reporting back and closing the issue 👍 :coffee: 

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
